### ISSUE TITLE
asynchronous setter interface

### ISSUE THREAD

**Issue by knksmith57** [expressjs/cookie-session#75]
Heya @dougwilson et al., reaching out seeking wisdom and guidance.

*tl;dr*: I need to perform an async operation after all session properties are set to compute and set a final value.

consider this simplified example:
* there are many discrete middleware handlers that may modify `req.session.*` properties
* before the response is sent, but after all possible middleware handlers have executed, 1 final _computed_ property needs to be set
* the computation of this property is an asynchronous operation

as I understand it, because [`on-headers`](https://github.com/jshttp/on-headers) is used [to perform the conditional `cookies.set` operation](https://github.com/expressjs/cookie-session/blob/34ac076c9809e0d5369e367baee484eac56e243c/index.js#L118), and `on-headers` has no callback interface, all operations must be synchronous. is that assumption correct?

right now I am handling this by manually patching `res.end`. it feels sketchy and brittle:
```js
app.use(function asyncCookieSetterMiddleware (req, res, next) {
  // stash original `end` method to call after we've handled our async work
  var originalEndFn = res.end
  
  // patch `res.end` to perform our cookie setting before ending the response
  res.end = function onResEnd (data, encoding, callback) {
    res.end = originalEndFn
  
    var shouldIssueNewCookie = someAppLogic()
  
    if (shouldIssueNewCookie) {
      someAsyncFunctionThatReturnsAPromise(req.session)
        .then(computedValue => {
          res.cookie('someProp', computedValue, { /* options */ })
          originalEndFn.call(res, data, encoding, callback)
        })
        .catch(next) // rewinds back to middleware registration point so that we fall into our error handling chain
    } else {
      originalEndFn.call(res, data, encoding, callback)
    }
  }
  
  next()
})
```

Y'all are the express masters. How should I _actually_ be doing this?

Thanks for your time and for the read!

**Comment by dougwilson** [comment]
Hi @knksmith57 thanks for the question :) I wanted to rely even before I had a full answer to (a) let you know I saw it and (b) at least put down some partial thoughts.

The biggest issue here is that you want to wait until the request is "over", and intercept it to perform an async action _but at the same time this needs to happen prior to the headers, since the entire session is stored in the header_. This is complicated. The `on-headers` is sync because that is how Node.js core is written; the interface has the expectation that when `res.writeHead` is called, the next line the fully-rendered headers are there and ready to write on the wire -- and of course this is the only generic "I'm done with the response" that is in Node.js core.

You're right that the `res.end` patch is brittle -- it doesn't even work with `stream.pipe(res)` or any other construct that either (a) calls a `res.write()` prior to `res.end()` or (b) calls `res.writeHead()` prior to `res.end()`. Even patching up `res.write()` is tricky because you have to create a buffer for code like `res.write(); res.end()`, with any number of writes.

Ideally this module would offer a hook you could use -- but fundamentally that hook would still be sync due to the Node.js core interface limitations.

I'm not, though, suggesting this is _unsolvable_, which is why I'm still thinking more on it, but just wanted to jot down some initial thoughts :) and perhaps add some more insight.

**Comment by dougwilson** [comment]
Yea, I agree with your assessment :/

**Comment by dougwilson** [comment]
I'm closing this because I don't think there is a way forward from the discussion. But maybe I missed it -- please feel free to let me know if I incorrectly closed :) !
